src/binary-encoding.adoc

```diff
@@ -0,0 +1,1618 @@
+= Binary Encoding
+:toc:
+
+[[encoding_primitives]]
+== Encoding primitives
+
+[[varuint]]
+=== `VarUInt`
+
+A variable-length unsigned integer.
+
+The bytes of a ``VarUInt``s are written in
+link:https://en.wikipedia.org/wiki/Endianness:[Little Endian byte order]. This means that the first byte will contain
+the ``VarUInt``'s least significant bits.
+
+The least significant bits in the `VarUInt` indicate the number of bytes that were used to encode the integer.
```

Comment:
Sorry for jumping in late and missing the convo here...

I assume we considered specifying the length prefix as the "first" bits, as opposed to the least significant in the first byte(s). That strikes me as cleaner/clearer _a priori_ but I'm guessing this was chosen after some _a posteriori_ observations that it is simpler/faster to implement as specified.

Reply:
Good question! Consider this example from the doc:
```
             ┌──── There's 1 zero in the least significant bits, so this
             │     integer is two bytes wide.
            ┌┴┐
0 1 1 0 0 1 1 0  0 0 0 0 1 0 1 1
└────┬────┘      └──────┬──────┘
lowest 6 bits    highest 8 bits
of the unsigned  of the unsigned
integer          integer
```

When the two bytes are read into an integer, the least significant bits of the first byte will be at the "end."

```
0 0 0 0 1 0 1 1  0 1 1 0 0 1 1 0
└──────┬──────┘  └────┬────┘ └┬┘
    highest 8      lowest 6   └── Continuation flag bits
      bits           bits
```

This enables a reader to examine the first byte in the serialized form and use a "count trailing zeros" operation (on some platforms this is "reverse bits" + "count leading zeros") to get a count of the bytes and read them into an integer. Then the entire integer can be shifted right by `num_bytes` (the number of flag bits) to get the value itself.

If instead we were to store the continuation flags in the "first" (highest) bits of the first byte(s) like this:
```
 ┌──── Continuation flag bits
┌┴┐
1 0 0 1 1 0 0 1  0 0 0 0 1 0 1 1
    └────┬────┘  └──────┬──────┘
lowest 6 bits    highest 8 bits
of the unsigned  of the unsigned
integer          integer
```
that would mean that the continuation bits would appear midway through the magnitude bits like this:
```
                  ┌──── Continuation flag bits
                 ┌┴┐
0 0 0 0 1 0 1 1  0 1 1 0 0 1 1 0
└──────┬──────┘      └────┬────┘ 
    highest 8          lowest 6   
      bits               bits
```

Reply:
Makes sense. Thanks!